### PR TITLE
[Feat] Add CUTLASS matmul-epilogue fusion path

### PR DESCRIPTION
**Pull request by wtr0504**
## 🗂️ PR Category

- [x] ✨ New Feature
- [x] 🚀 Optimization (performance, memory, etc.)
- [ ] 💥 Breaking Change
- [ ] 🐛 Bug Fix
- [ ] 🛠️ Development / Refactoring
- [ ] 📚 Documentation
- [ ] 🧹 Chore (Dependencies, CI/CD, Configuration, etc.)
- [ ] 🧪 Testing

## 📝 Description
 Add a CUTLASS-based matmul + epilogue fusion pass that fuses aten.mm followed by elementwise chains (activations, scalar ops, bias-add, residual loads) into a single GPU kernel, eliminating intermediate global memory round-trips. 
                                                                         
### Two fusion backends

- **Generic EVT (Epilogue Visitor Tree)** — builds an IR tree from the FX epilogue chain and JIT-compiles a CUTLASS kernel templated on that tree. Supports unary activations (SiLU, Sigmoid, GeLU, ReLU, etc.), scalar arithmetic, 1-D bias (RowBroadcast), column scaling (ColBroadcast), and full (M, N) auxiliary loads. Codegen renders to CUTLASS 3.x `Sm90EVT` on H100 (sm_90) and CUTLASS 2.x `Sm80EVT` on RTX 5090 / Blackwell consumer (sm_120).

- **SwiGLU DualGemm** — pattern-matches the canonical SwiGLU recipe (`slice-stride-2 → dual clamp → scaled SiLU → multiply`) and dispatches to a vendored CUTLASS `DualGemm` kernel that runs both GEMMs (`A @ W_gate.T` and `A @ W_linear.T`) in the same threadblock, sharing A's SMEM stages. Writes `(M, N/2)` directly. Routes to SM80 `cp.async` multistage on sm_120 and to SM90 TMA + WGMMA warp-specialized path on sm_90. Static constants (`alpha`, `limit`, `one`) are captured dynamically from the FX graph.                                                              
                        
### Key design points

- **Per-node `compute_dtype`**: the IR walker tracks `_to_copy` / `convert_element_type` nodes and stamps each `Compute` node with the precision active at that point (`float32`, `bfloat16`, or `float16`). Codegen emits per-node element types in `VisitorCompute` / `Sm90Compute`.
- **Greedy alignment**: tries 128-bit loads first, falls back to 64-bit when K or N only meets 8-byte alignment. The runtime pads D's row stride to 16-byte boundaries for TMA compatibility.
- **Autotune**: both DualGemm paths register multiple (TileShape, Stages) candidates and time them at first call per shape bucket; the winner is cached for all subsequent calls.
- **Build robustness**: `_track_build` / `_untrack_build` + `atexit` + SIGTERM/SIGINT/SIGHUP handlers ensure interrupted `cpp_extension.load` calls don't leave stale lock files. Warm-cache fast path (`_try_dlopen_prebuilt`) skips `cpp_extension.load` entirely when the `.so` already exists, preventing multi-rank FileBaton hangs.
- **SM90 multi-AuxLoad**: multiple `(M, N)` auxiliary tensors are loaded via inline `ld.global` (`Sm90AuxLoad<0>`) — no SMEM staging needed — enabling patterns like `mm + R1 + R2 + R3`.                                                                                                 
 
### Files added / changed

| Area | Files | What |
|------|-------|------|
| IR | `evt_ir.py` | Dataclass IR (Accum, Compute, Store, RowBroadcast, ColBroadcast, AuxLoad) + canonical JSON serialization |
| FX pass | `matmul_epilogue_fusion.py` | Graph walker, epilogue chain absorption, SwiGLU pattern matching, B-layout classification |
| Runtime | `evt_runtime.py` | `torch.library` op, JIT compile cache, DualGemm loader, dispatch fast-cache, build cleanup |
| SM80 codegen | `sm80/evt_codegen.py` | CUTLASS 2.x Sm80EVT `.cu` renderer |
| SM90 codegen | `sm90/evt_codegen.py` | CUTLASS 3.x Sm90EVT `.cu` renderer with Sm90Compute + Sm90AuxLoad |
| SM80 DualGemm | `sm80/cutlass_kernels/swiglu_one_stage.cu` | Vendored DualGemm + SwigluCombine epilogue + autotune runner |
| SM90 DualGemm | `sm90/cutlass_kernels/swiglu_one_stage.cu` | TMA + WGMMA DualGemm + SwigluCombine + autotune runner |
| SM90 device wrapper | `sm90/cutlass_kernels/hopper_dual_gemm/` | Vendored Sm90DualGemm with LayoutTraits for 2.x→3.x layout translation |
| Shared | `common/cutlass_kernels/swiglu_combine.h`, `common/codegen_shared.py` | SwigluCombine functor, shared codegen utilities |
| Config | `config.py` | `enable_mm_epilogue_fusion` flag in PassConfig |
| Infra | `Dockerfile`, `.pre-commit-config.yaml` | CUTLASS install, copyright hook |
| Tests | `test_matmul_epilogue_fusion.py` | 30+ tests: positive (activations, scalar ops, bias, AuxLoad, SwiGLU), negative (escape, misalign, bare mm), out_dtype matrix, compute_dtype, D-stride padding, SM90 parity |
| Tests | `test_build_cleanup.py` | 5 tests for the `_track_build`/`_untrack_build` + signal-handler cleanup mechanism |

### CODE REVIEW

File: magi_compiler/passes/full_graph/remove_useless_ops.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+# limitations under the License.
+
+import torch
+import torch._inductor.fx_passes.pre_grad
+
+from ...magi_depyf.timeline import emit_pass_lifecycle
+from ..pass_base import MagiInductorPass
+
+
+class RemoveUselessOpsPass(MagiInductorPass):
```
Comment (jiahy0825) [inline magi_compiler/passes/full_graph/remove_useless_ops.py]: `RemoveUselessOpsPass` -> `EliminateIdentityViewCastPass`
Note: also change the file name

File: magi_compiler/utils/device.py
Comment (jiahy0825) [inline magi_compiler/utils/device.py:1]: Move this file to the `utils` directory and update `__init__.py`

File: Dockerfile [hunk 2]
```diff
@@ -3,6 +3,21 @@ FROM nvcr.io/nvidia/pytorch:25.10-py3
 
 ARG FLASH_ATTENTION_COMMIT_ID="b613d9e2c8475945baff3fd68f2030af1b890acf"
 
+# CUTLASS — source is always cloned (the magi_compiler EVT-fusion path
```
Comment (jiahy0825) [inline Dockerfile:6]: Treat cutlass as third-party and provide install cmd cause users may install magi_compiler without docker.
Update commands in readme.md pls~

File: magi_compiler/passes/piecewise_graph/post_grad_pass_manager.py [hunk 3]
```diff
@@ -80,7 +82,13 @@ def __call__(self, graph: fx.Graph):
     def configure(self, pass_config: PassConfig):
         self.pass_config = pass_config
 
-        # TODO: Register custom passes here (fusion, noop elimination, sequence parallelism, async TP, Ulysses overlap).
+        # Matmul + epilogue fusion. On sm_120 (Blackwell consumer / RTX 5090)
+        # we lower fused chains to a CUTLASS Sm80EVT kernel. Toggled via
+        # PassConfig.enable_mm_epilogue_fusion (default True). The device
+        # check is independent — even with the flag on, non-sm_120 hosts
+        # don't register the pass since its FX walker would just no-op.
```
Comment (jiahy0825) [inline magi_compiler/passes/piecewise_graph/post_grad_pass_manager.py]: It seems better to support epilogue fusion for other arches, even if it achieves no performance gain.
Cause this fusion can also save memory.

File: magi_compiler/passes/piecewise_graph/fusion/blackwell_geforce/matmul_epilogue_fusion.py [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+                args=(A, b_underlying, extras_nodes, ir_json, kind, n_out, out_dt_id),
+            )
+        # Propagate FakeTensor meta so downstream Inductor checks pass.
+        try:
+            val_last = last_node.meta.get("val")
+            if val_last is not None:
+                # Propagate but with 128B-aligned stride matching what the
+                # CUDA impl actually returns.
+                new_val = val_last.new_empty_strided(
+                    val_last.shape, (evt_runtime._aligned_n_stride(int(val_last.shape[-1]), val_last.dtype), 1)
```
Comment (jiahy0825) [inline magi_compiler/passes/piecewise_graph/fusion/blackwell_geforce/matmul_epilogue_fusion.py]: `_aligned_n_stride` not defined

File: magi_compiler/passes/piecewise_graph/fusion/blackwell_geforce/evt_runtime.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+        src_path = os.path.join(build_dir, "evt.cu")
+        # Write atomically (tmp + rename) so concurrent processes don't see a
+        # half-written file. Use a process-specific tmp name to avoid races
+        # across multiple rank processes generating the same kernel.
+        tmp_path = f"{src_path}.{os.getpid()}.tmp"
+        with open(tmp_path, "w") as f:
+            f.write(src)
+        os.replace(tmp_path, src_path)
+
+        cutlass_root = _cutlass_root()
```
Comment (jiahy0825) [inline magi_compiler/passes/piecewise_graph/fusion/blackwell_geforce/evt_runtime.py]: Find a better way to find the cutlass_root dir~

File: magi_compiler/passes/piecewise_graph/fusion/matmul_epilogue_fusion.py [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+                curr = curr.next
+                continue
+
+            # ── Type conversion (no-op in fp32 EVT) ───────────────────────────
+            if target in _TYPE_CONV_OPS:
+                node_to_ir[curr] = node_to_ir[curr.args[0]]
+                walk_seen.append(curr)
+                last_node = curr
+                last_ir = node_to_ir[curr]
+                curr = curr.next
```
Comment (jiahy0825) [inline magi_compiler/passes/piecewise_graph/fusion/matmul_epilogue_fusion.py]: If we have a compute chain like this:
`matmul -> to_1(fp32) -> func1 -> to_2(fp8) -> func2`
What happens to type conversion ops like to_1 and to_2?

Comment (wtr0504) [reply magi_compiler/passes/piecewise_graph/fusion/matmul_epilogue_fusion.py]: Type conversion ops (to_1, to_2) in the EVT fusion chain:
  
  Both convert_element_type and _to_copy are absorbed by the walker — they update current_compute_dtype so that subsequent Compute nodes are instantiated at the target precision, but no
  separate IR node is emitted for the cast itself (CUTLASS VisitorCompute handles the implicit type conversion between its ElementInput and ElementOutput template parameters).

  This example: matmul → to_1(fp32) → func1 → to_2(fp8) → func2

  - to_1(fp32) sets current_compute_dtype = "float32" → func1 runs in fp32.
  - to_2(fp8) — fp8 is not in _DTYPE_TO_STR (which only maps {bf16, fp16, fp32}), so current_compute_dtype stays "float32". func2 still runs in fp32. The chain continues fusing but fp8 is
  silently ignored as a compute precision.

  Currently supported conversions: only float32, float16, and bfloat16 — matching the CUTLASS VisitorCompute / Sm90Compute template-supported types.

**Review by jiahy0825 — APPROVED** [review]
LGTM
